### PR TITLE
feat: limit size of temp cache folder

### PR DESCRIPTION
**Pull request by suprko**
[PR guideline](https://github.com/PeerioTechnologies/peerio-icebear/blob/dev/docs/CONTRIBUTING.md)

#### Relevant issue tracker links  
https://peerio.tpondemand.com/entity/7067
https://app.clubhouse.io/peerio/story/277/desktop-limit-file-system-cache

#### Linked PRs in other repos
desktop https://github.com/PeerioTechnologies/peerio-desktop/pull/167
#### Breaking changes  
n/a
#### Manual testing instructions  
On this branch this app should:
1. (unchanged) Download inline images to temp cache folder
2. When folder is over an arbitrary size limit, delete temp images (oldest first) until folder is below that limit, *except* -
3. Never delete temp images that are in the currently selected chat

#### Additional notes
n/a

----
### Reviewer

* Localisation.
* Architectural decisions.
* Performance.
* Comments.
* Big feature tech design documentation.
* Readability.
* DRY.
* Are there any bicycles invented that should be replaced with existing code or external libs?
* Are there unit and e2e tests added?

----
### Repository owner

- [ ] Breaking changes are applied correctly.
- [ ] At least one approved code review exists.
- [ ] Approved tester review exists and was done after last code changes.
- [ ] CI build and tests passed.
- [ ] This feature/bugfix is considered fully done and can get merged.
- [ ] PR name follows conventional changelog format.
- [ ] I'll remember to squash commits.

### CODE REVIEW

**Comment by seavan** [comment]
waiting for @anri-asaturov

**Comment by flohdot** [comment]
Hmmm, this is a bit hard to understand (for me) because it's split across files, and I have a feeling someone is gonna be confused by it later on. Can we get some inline comments that explain what's up?

For example, in client-app.js, it would help to have something that tells me why this inlineFileVisible stuff is there.

**Comment by suprko** [comment]
@flohdot added some comments

File: src/models/client-app.js [hunk 1]
```diff
@@ -106,6 +106,20 @@ class ClientApp {
...
+     * @public
+     */
+    inlineFilesVisible = {};
+
+    addFileVisible(fileId) {
+        this.inlineFilesVisible[fileId] = true;
+    }
+
+    removeFileVisible(fileId) {
+        this.inlineFilesVisible[fileId] = false;
```
Comment (lostfictions) [inline src/models/client-app.js]: This smells a bit like a leak -- wouldn't you just want to `delete` the key instead? (Or maybe even better, just use a `Set`?)

File: src/models/files/file-cache-handler.js [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+     * deleting files (oldest first) that are not currently visible
+     */
+    static cacheDeleteLoop() {
+        for (let i = 0; i < files.length && fileSizeTotal > fileSizeLimit; i++) {
+            const f = files[i];
+            if (clientApp.inlineFilesVisible[files[i].fileId]) continue;
+            config.FileStream.delete(f.tmpCachePath);
+            f.tmpCached = false;
+            fileSizeTotal -= f.size;
+            files.splice(i--, 1);
```
Comment (lostfictions) [inline src/models/files/file-cache-handler.js]: This is a bit too clever to me. There's very, very few valid uses of the prefix increment/decrement operators inline in another statement, and slim-to-none uses of the postfix operators inline. If you're performing two mutations on two variables -- and especially if there's a dependent ordering to those two mutations -- I'd recommend you use two statements for clarity.

Comment (lostfictions) [reply src/models/files/file-cache-handler.js]: incidentally, `splice` is one of the slowest array methods... it's O(n), and it's possibly being called here in a way that's towards the worst-case of n here. plus it's potentially being called for every item in the array, making this O(n^2), every time an image is added to the cache. has this been profiled at all? i could definitely see it causing noticeable extra jank. i also wonder if a proper LRU could make this a lot simpler.

Comment (seavan) [reply src/models/files/file-cache-handler.js]: @lostfictions yup you're right, no place for splice here

**Comment by seavan** [comment]
after discussion with @lostfictions confirmed https://github.com/rsms/js-lru as a no-option - badly maintained leaky js piece of code.

going with node-lru-cache. refactoring.

File: src/config.js [hunk 3]
```diff
@@ -338,6 +338,11 @@ const config = new class {
     };
 
     unfurlTimeout = 30000;
+
+    /**
+     * Maximum total number of cached images which we store before we start deleting the least recent ones
```
Comment (lostfictions) [inline src/config.js]: number of images?

File: src/models/files/file-cache-handler.js [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+    length(file) {
+        return file.size;
+    },
+    // remove the item from cache
+    dispose(key, file) {
+        // to avoid solving probable but unlikely complicated situations when total size of visible files
+        // exceeds cache limit, we are deleting file whenever it does become invisible
+        // it may lead to actual cache size exceeding our limit, but this should be a non-issue
+        console.debug(`trying to dispose: ${key}, ${file.tmpCachePath}`);
+        when(() => !file.visibleCounter, () => {
```
Comment (anri-asaturov) [inline src/models/files/file-cache-handler.js]: nice idea

File: src/models/files/file.download.js [hunk 5]
```diff
@@ -117,12 +119,32 @@ function cancelDownload() {
...
+ * @returns {Promise}
+ * @instance
+ * @memberof File
+ * @public
+ */
+function removeCache() {
+    return Promise.resolve((async () => {
+        if (!this.tmpCached) return;
+        try {
+            await config.FileStream.delete(this.tmpCachePath);
```
Comment (anri-asaturov) [inline src/models/files/file.download.js]: I think delete() will reject if file is already deleted for whatever reason, maybe we should reset tmpCached flag in any case?

**Comment by seavan** [comment]
@ebilg ready for testing

File: src/models/files/file.download.js (hunk 5 above)
Comment (seavan) [reply src/models/files/file.download.js]: done

**Pull request merged**